accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+# Function Pointers
+#### Steve Harter
+#### December 13, 2022
+
+# Background
+For additional context, see:
+- The [original feature ask](https://github.com/dotnet/runtime/issues/11354)
+- The [MetadataLoadContext feature ask](https://github.com/dotnet/runtime/issues/43791)
+- The [7.0 introspection attempt](https://github.com/dotnet/runtime/pull/71516)
+- [C# function pointer spec](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers)
+
+Function pointers, which have been in ECMA-335 and used by the CLR since the beginning were only recently supported by C# in C# v9.0 ad .NET v6. However, the .NET type system has not been updated to directly support them and they are currently exposed as the `IntPtr` type:
+```cs
+// using typeof:
+Type type = typeof(delegate*<int, bool>); // System.IntPtr (currently)
+
+// using Type.GetType():
+delegate*<int, bool> fn = &MyFunctionPointer;
+type = fn.GetType(); // System.IntPtr (currently)
+
+static bool MyFunctionPointer(int n) { return n == 42; }
+```
+
+Reflection also returns `IntPtr` for both runtime-reflection through `FieldInfo\ParameterInfo\PropertyInfo` and static-reflection through `MetadataLoadContext`:
+```cs
+  FieldInfo fi = typeof(MyClass).GetField("_fn");
+  Type type = fi.FieldType; // System.IntPtr (currently)
+
+  unsafe class MyClass
+  {
+      public delegate*<int, bool> _fn; // A field as a function pointer
+  }
+```
+
+Treating function pointers as `IntPtr` essentially makes reflection introspection impossible, blocking scenarios that care about function pointers. The proposal is to change from `IntPtr` to `Type` which is a **breaking change**:
```

Comment:
I have a scenario this spec would hopefully improve:

I want to create a method that takes a function pointer (any function pointer), and then reflects on the signature of that function pointer. Right now that's not possible, and the only way I could come up with to implement this was to add another parameter to the function that takes a MethodInfo for the target of the function pointer:

```cs
void DoSomething (void *fptr, MethodInfo method);
```

While it works, the calling code is quite ugly.

In order to take advantage of the new reflection capability, I guess I'd have to type the function pointer as object?

```cs
void DoSomething (object fptr)
{
    if (!ftpr.GetType ().IsFunctionPointer)
        throw new InvalidArgumentException ("Only function pointers!");
}
```

Unfortunately the calling code is not very nice on the eyes:

```cs

static void Y () { }

DoSomething (&Y); // doesn't work: error CS1503: Argument 1: cannot convert from '&method group' to 'object'
DoSomething ((delegate*<void>)&Y); // doesn't work either: error CS1503: Argument 1: cannot convert from 'delegate*<void>' to 'object'
DoSomething ((object)(delegate*<void>)&Y); // nope: error CS1503: Argument 1: cannot convert from 'delegate*<void>' to 'object'
DoSomething ((IntPtr)(delegate*<void>)&Y); // this works, but it's not very intuitive
DoSomething ((object)(IntPtr)(delegate*<void>)&Y); // this works, and it's also obvious why it works, but seriously, _3_ casts?
```

I thought about using generics, but that [doesn't work](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEAMACdBGAOgCUBXGAFyjAFMCBJC6gJwHsAHAZWYDcoBjagGcA3AFgAUBPQAmbAEE2bCQG8J2dbgDMuPADZcAFmwARFhxY1yAC1gBzADwAVAHzYAFI+wAzNuSYBKNQ1VcQ1sAF8g9XRtfH10IwBNd0DQ4KiIjJjsMkEIL2odeKMAWWgYd3xMAG0AXWwIJltBVLCQsLDTc0sbGAcEagAbalsIcmoAKnsE51c3ADJE/zE09UjxcKA):

```cs
void DoSomething<T> (T fptr) { }
void Y () { }
unsafe void F ()
{
    DoSomething<delegate*<void>> (&Y);
}
// error CS0306: The type 'delegate*<void>' may not be used as a type argument
```

Using a `System.Reflection.FunctionPointer` type as suggested at the end of this spec won't work (even if it were implemented), because there's no API to get the type of the function pointer.

What I'm hoping for is this to work somehow:

```cs
void DoSomething (FunctionPointer fptr) {}
static void Y () {}
void F ()
{
    DoSomething (&Y);
}
```

This would obviously require changes to both the runtime and the C# compiler, but it would be nice if the runtime part could be designed so that at least it's possible to implement the C# bits.

Reply:
> I want to create a method that takes a function pointer (any function pointer), and then reflects on the signature of that function pointer.

The polymorphic limitations here should be the same as dealing with a standard pointer - a pointer instance isn't boxable and thus you can't call `GetType()` on it:
```cs
int i = 1;
int* ip = &i;
ip.GetType(); // CS1061
```

Do you need to call the function pointer? If not, just passing the `Type` should be fine:

> void DoSomething (void *fptr, MethodInfo method);

You can modify this and just pass `Type` instead of `MethodInfo`:
```cs
static void Y() { }

delegate*<void> y = &Y;
DoSomething(y, y.GetType());
   
// Unless we want to call it, fnPtr is not required
void DoSomething(void* fnPtr, Type fnType)
{
    if (!fnType.IsFunctionPointer) throw new ArgumentException...
    
    // Inspect fnType here - parameter types, etc
}
```
Also in the "future apis" section there is a "MethodSignature" class that exposes the metadata from a function pointer and may also have late-bound Invoke capabilities. Not sure if that would help with your scenario.

Reply:
> I want to create a method that takes a function pointer (any function pointer), and then reflects on the signature of that function pointer.

The polymorphic limitations here should be the same as dealing with a standard pointer - a pointer instance isn't boxable and thus you can't call `GetType()` on it:
```cs
int i = 1;
int* ip = &i;
ip.GetType(); // CS1061
```

Do you need to call the function pointer? If not, just passing the `Type` should be fine:

> void DoSomething (void *fptr, MethodInfo method);

You can modify this and just pass `Type` instead of `MethodInfo`:
```cs
static void Y() { }

delegate*<void> y = &Y;
DoSomething(y, typeof(delegate*<void>));
   
// Unless we want to call it, fnPtr is not required
void DoSomething(void* fnPtr, Type fnType)
{
    if (!fnType.IsFunctionPointer) throw new ArgumentException...
    
    // Inspect fnType here - parameter types, etc
}
```
Also in the "future apis" section there is a "MethodSignature" class that exposes the metadata from a function pointer and may also have late-bound Invoke capabilities. Not sure if that would help with your scenario.

Reply:
> You can modify this and just pass Type instead of MethodInfo:

Yes, that looks like it could be an improvement (avoids reflection at the callsite at least).

Thanks!